File: reverse-v2/ghidra-scripts/ExportDecomp.java
// Ghidra Java postScript: export decompiled C for all functions + metadata.
// Args: <outdir>
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.*;
import ghidra.program.model.listing.*;
import ghidra.program.model.symbol.*;
import ghidra.program.model.mem.MemoryBlock;
import ghidra.program.model.listing.Data;
import java.io.*;
import java.util.*;

public class ExportDecomp extends GhidraScript {
    public void run() throws Exception {
        String[] args = getScriptArgs();
        String outdir = args.length > 0 ? args[0] : "/out";
        new File(outdir).mkdirs();

        Program prog = currentProgram;
        FunctionManager fm = prog.getFunctionManager();

        DecompInterface di = new DecompInterface();
        DecompileOptions opts = new DecompileOptions();
        di.setOptions(opts);
        di.toggleCCode(true);
        di.toggleSyntaxTree(true);
        di.setSimplificationStyle("decompile");
        di.openProgram(prog);

        List<Function> funcs = new ArrayList<>();
        for (Function f : fm.getFunctions(true)) funcs.add(f);
        println("[export] functions: " + funcs.size());

        BufferedWriter cAll = new BufferedWriter(new FileWriter(new File(outdir, "decompiled_all.c")));
        BufferedWriter idx = new BufferedWriter(new FileWriter(new File(outdir, "functions_index.csv")));
        idx.write("addr,name,size,decomp_ok\n");

        int ok = 0, fail = 0, i = 0;
        for (Function f : funcs) {
            i++;
            String name = f.getName();
            String ep = f.getEntryPoint().toString();
            long size = f.getBody().getNumAddresses();
            DecompileResults res = null;
            boolean good = false;
            try {
                res = di.decompileFunction(f, 60, monitor);
                good = res != null && res.decompileCompleted();
            } catch (Exception e) { good = false; }
            idx.write(ep + "," + name.replace(",", "_") + "," + size + "," + (good ? 1 : 0) + "\n");
            cAll.write("\n/* ==== " + name + " @ " + ep + "  size=" + size + " ==== */\n");
            if (good) {
                try { cAll.write(res.getDecompiledFunction().getC()); ok++; }
                catch (Exception e) { cAll.write("/* getC failed */\n"); fail++; }
            } else { cAll.write("/* decompile failed */\n"); fail++; }
            if (i % 200 == 0) println("[export] " + i + "/" + funcs.size() + " ok=" + ok + " fail=" + fail);
        }
        cAll.close(); idx.close();
        println("[export] decompiled ok=" + ok + " fail=" + fail);

        // symbols
        BufferedWriter sym = new BufferedWriter(new FileWriter(new File(outdir, "symbols.txt")));
        for (Symbol s : prog.getSymbolTable().getAllSymbols(true)) {
            try { sym.write(s.getAddress() + " " + s.getSymbolType() + " " + s.getName() + "\n"); } catch (Exception e) {}
        }
        sym.close();

        // strings (defined data whose value is a String)
        BufferedWriter str = new BufferedWriter(new FileWriter(new File(outdir, "strings.txt")));
        try {
            for (Data d : prog.getListing().getDefinedData(true)) {
                try {
                    Object val = d.getValue();
                    if (val instanceof String) {
                        str.write(d.getAddress() + "\t" + ((String) val).replace("\n", " ").replace("\r", " ") + "\n");
                    }
                } catch (Exception e) {}
            }
        } catch (Exception e) { println("[export] strings err " + e.getMessage()); }
        str.close();

        // memory map
        BufferedWriter mm = new BufferedWriter(new FileWriter(new File(outdir, "memory_map.txt")));
        for (MemoryBlock b : prog.getMemory().getBlocks()) {
            mm.write(b.getStart() + " - " + b.getEnd() + "  " + b.getName()
                + "  r=" + b.isRead() + " w=" + b.isWrite() + " x=" + b.isExecute() + " vol=" + b.isVolatile() + "\n");
        }
        mm.close();
        println("[export] DONE -> " + outdir);
    }
}

File: reverse-v2/ghidra-scripts/export_decomp.py
# Ghidra Jython headless postScript: export decompiled C for all functions + metadata.
# Usage: -postScript export_decomp.py <outdir>
import os
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

args = getScriptArgs()
outdir = args[0] if args else "/out"
if not os.path.exists(outdir):
    os.makedirs(outdir)

prog = currentProgram
fm = prog.getFunctionManager()
monitor = ConsoleTaskMonitor()

decomp = DecompInterface()
decomp.toggleCCode(True)
decomp.toggleSyntaxTree(True)
decomp.setSimplificationStyle("decompile")
decomp.openProgram(prog)

funcs = list(fm.getFunctions(True))
print("[export] %d functions" % len(funcs))

# 1) combined decompiled C + per-function index
c_all = open(os.path.join(outdir, "decompiled_all.c"), "w")
idx = open(os.path.join(outdir, "functions_index.csv"), "w")
idx.write("addr,name,size,decomp_ok\n")

ok = 0
fail = 0
for i, f in enumerate(funcs):
    ep = f.getEntryPoint()
    name = f.getName()
    size = f.getBody().getNumAddresses()
    res = None
    try:
        res = decomp.decompileFunction(f, 60, monitor)
    except Exception as e:
        res = None
    good = res is not None and res.decompileCompleted()
    idx.write("%s,%s,%d,%d\n" % (ep, name, size, 1 if good else 0))
    c_all.write("\n/* ==== %s @ %s  size=%d ==== */\n" % (name, ep, size))
    if good:
        try:
            c_all.write(res.getDecompiledFunction().getC())
            ok += 1
        except:
            c_all.write("/* getC failed */\n")
            fail += 1
    else:
        c_all.write("/* decompile failed */\n")
        fail += 1
    if i % 200 == 0:
        print("[export] %d/%d (ok=%d fail=%d)" % (i, len(funcs), ok, fail))

c_all.close()
idx.close()
print("[export] decompiled ok=%d fail=%d" % (ok, fail))

# 2) symbols
st = prog.getSymbolTable()
fsym = open(os.path.join(outdir, "symbols.txt"), "w")
for s in st.getAllSymbols(True):
    try:
        fsym.write("%s %s %s\n" % (s.getAddress(), s.getSymbolType(), s.getName()))
    except:
        pass
fsym.close()

# 3) strings (defined data of string type)
from ghidra.program.util import DefinedDataIterator
fstr = open(os.path.join(outdir, "strings.txt"), "w")
try:
    for d in DefinedDataIterator.definedStrings(prog):
        try:
            fstr.write("%s\t%s\n" % (d.getAddress(), d.getValue()))
        except:
            pass
except Exception as e:
    print("[export] strings err %s" % e)
fstr.close()

# 4) memory map
fmm = open(os.path.join(outdir, "memory_map.txt"), "w")
for b in prog.getMemory().getBlocks():
    fmm.write("%s - %s  %s  r=%s w=%s x=%s vol=%s\n" % (
        b.getStart(), b.getEnd(), b.getName(),
        b.isRead(), b.isWrite(), b.isExecute(), b.isVolatile()))
fmm.close()

print("[export] DONE -> %s" % outdir)
